File: apps/screen/app/schemas/translation.py
"""Pydantic schemas for translation endpoints."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class LanguageResponse(BaseModel):
    """Language information response."""

    code: str = Field(..., description="ISO language code (e.g., 'fr', 'es')")
    name: str = Field(..., description="Language name in English")

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {"code": "fr", "name": "French"},
            ],
        },
    )


class LanguageTranslationStatus(BaseModel):
    """Translation status for a single language."""

    language_name: str = Field(..., description="Language name in English")
    status: str = Field(..., description="Translation status: 'complete', 'partial', or 'none'")
    fields_translated: int = Field(..., description="Number of translated fields")
    fields_total: int = Field(..., description="Total translatable fields")
    percentage: float = Field(..., description="Translation completion percentage")
    active_job: "TranslationJobResponse | None" = Field(None, description="Active translation job if any")


class CompanyTranslationStatusResponse(BaseModel):
    """Translation status response for a company."""

    company_id: int = Field(..., description="Company ID")
    translations: dict[str, LanguageTranslationStatus] = Field(..., description="Translation status per language code")

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "company_id": 42,
                    "translations": {
                        "fr": {
                            "language_name": "French",
                            "status": "complete",
                            "fields_translated": 15,
                            "fields_total": 15,
                            "percentage": 100.0,
                            "active_job": None,
                        },
                        "es": {
                            "language_name": "Spanish",
                            "status": "none",
                            "fields_translated": 0,
                            "fields_total": 15,
                            "percentage": 0.0,
                            "active_job": None,
                        },
                    },
                },
            ],
        },
    )


class TranslateRequest(BaseModel):
    """Request to translate company fields to a language."""

    language_code: str = Field(
        ...,
        description="Target language code (e.g., 'es', 'de', 'pt')",
        min_length=2,
        max_length=5,
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {"language_code": "fr"},
            ],
        },
    )


class TranslationJobResponse(BaseModel):
    """Translation job information."""

    id: int = Field(..., description="Job ID")
    company_id: int = Field(..., description="Company ID")
    language_code: str = Field(..., description="Target language code")
    status: str = Field(..., description="Job status: pending, running, completed, failed")
    total_fields: int = Field(..., description="Total fields to translate")
    translated_fields: int = Field(..., description="Fields translated so far")
    progress_percentage: float = Field(..., description="Progress percentage")
    error_message: str | None = Field(None, description="Error message if failed")
    created_at: datetime = Field(..., description="Job creation time")
    started_at: datetime | None = Field(None, description="Job start time")
    completed_at: datetime | None = Field(None, description="Job completion time")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "examples": [
                {
                    "id": 7,
                    "company_id": 42,
                    "language_code": "fr",
                    "status": "completed",
                    "total_fields": 15,
                    "translated_fields": 15,
                    "progress_percentage": 100.0,
                    "error_message": None,
                    "created_at": "2025-06-15T10:30:00Z",
                    "started_at": "2025-06-15T10:30:01Z",
                    "completed_at": "2025-06-15T10:30:03Z",
                },
            ],
        },
    )


class TranslateResponse(BaseModel):
    """Response after requesting translation."""

    company_id: int = Field(..., description="Company ID")
    language_code: str = Field(..., description="Target language code")
    fields_queued: int = Field(..., description="Number of fields queued for translation")
    message: str = Field(..., description="Status message")
    job: TranslationJobResponse | None = Field(None, description="Translation job details")

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "company_id": 42,
                    "language_code": "fr",
                    "fields_queued": 15,
                    "message": "Translation started for 15 fields.",
                    "job": {
                        "id": 7,
                        "company_id": 42,
                        "language_code": "fr",
                        "status": "pending",
                        "total_fields": 15,
                        "translated_fields": 0,
                        "progress_percentage": 0.0,
                        "error_message": None,
                        "created_at": "2025-06-15T10:30:00Z",
                        "started_at": None,
                        "completed_at": None,
                    },
                },
            ],
        },
    )


# Update forward reference
LanguageTranslationStatus.model_rebuild()

File: apps/front/src/api/translation.ts
/**
 * API functions for translation management.
 */

import { apiClient } from './client'

export interface TranslationLanguage {
  code: string
  name: string
}

export interface TranslationJob {
  id: number
  company_id: number
  language_code: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  total_fields: number
  translated_fields: number
  progress_percentage: number
  error_message: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
}

export interface LanguageTranslationStatus {
  language_name: string
  status: 'complete' | 'partial' | 'none'
  fields_translated: number
  fields_total: number
  percentage: number
  active_job: TranslationJob | null
}

export interface CompanyTranslationStatus {
  company_id: number
  translations: Record<string, LanguageTranslationStatus>
}

export interface TranslateResponse {
  company_id: number
  language_code: string
  fields_queued: number
  message: string
  job: TranslationJob | null
}

/**
 * Get available translation languages.
 *
 * @returns List of available language codes and names
 */
export async function getTranslationLanguages(): Promise<TranslationLanguage[]> {
  return apiClient.get<TranslationLanguage[]>('/translation/list')
}

/**
 * Get translation status for a company.
 *
 * @param companyId - Company ID to check status for
 * @returns Translation status for all languages
 */
export async function getCompanyTranslationStatus(
  companyId: number | string,
): Promise<CompanyTranslationStatus> {
  return apiClient.get<CompanyTranslationStatus>(`/translation/status/${companyId}`)
}

/**
 * Request translation of a company to a specific language.
 *
 * @param companyId - Company ID to translate
 * @param languageCode - Target language code
 * @returns Translation request response with job details
 */
export async function requestTranslation(
  companyId: number | string,
  languageCode: string,
): Promise<TranslateResponse> {
  return apiClient.post<TranslateResponse>(
    `/translation/translate/${companyId}`,
    {
      language_code: languageCode,
    },
    { silent: true },
  )
}

/**
 * Get translation job progress.
 *
 * @param jobId - Translation job ID
 * @returns Translation job details
 */
export async function getTranslationJob(jobId: number): Promise<TranslationJob> {
  return apiClient.get<TranslationJob>(`/translation/job/${jobId}`)
}
